library/pkcs7.c

```diff
@@ -574,6 +594,12 @@ int mbedtls_pkcs7_parse_der(mbedtls_pkcs7 *pkcs7, const unsigned char *buf,
         goto out;
     }
 
+    /* ensure no extra/missing data */
+    if (p + len != end) {
+        ret = MBEDTLS_ERR_PKCS7_BAD_INPUT_DATA;
+        goto out;
+    }
+
```

Comment:
At line 663 / 668, we have untested error paths for an unrecognised digest algorithm. It looks like these should be straight-forward to add test-cases for by adding PCKS7 files using disabled and invalid digest algs and using the existing pkcs7_parse test case.

Reply:
Agree would be nice to test this. Unfortunately, this block is in the verify function, so we would have to first have it pass the parsing function to get the pkcs7 struct. In parsing logic, unsupported/invalid digest algs are being caught by `pkcs7_get_signed_data`:
```
ret = pkcs7_get_digest_algorithm_set(&p, end_set,
                                         &signed_data->digest_alg_identifiers);
    if (ret != 0) {
        return ret;
    }

    ret = mbedtls_oid_get_md_alg(&signed_data->digest_alg_identifiers, &md_alg);
    if (ret != 0) {
        return MBEDTLS_ERR_PKCS7_INVALID_ALG;
    }
```
But   ` return MBEDTLS_ERR_PKCS7_INVALID_ALG` was not getting covered here either. So I will add a test for this.

Reply:
If this is genuinely unreachable, should there be an assertion?

Reply:
It could be reached by manually modifying fields in the pkcs7 structure I guess (indeed, this would be the way to test it). And in general, Mbed TLS does not use assertions. So I think the approach is correct.